### ISSUE TITLE
Trends with daily granularity

### ISSUE THREAD

**Issue by bigme666** [GeneralMills/pytrends#174]
I think that this is not an issue because Google decides the granularity of the results (daily, weekly or monthly) depending on the search time frame. So I decided to implement a method that splits big time frames into smaller ones (90 days i.e.) with a one day overlap to normalize the scale between the data.
Do you think this could be a good improvement for pytrends?

**Comment by dreyco676** [comment]
That would be an amazing extension. I know I was planning on doing
something like it a while ago but never had the time or need to do so.

**Comment by IljaK91** [comment]
I would really like it!

**Comment by bigme666** [comment]
I'm actually working on this. Stay tuned...

**Comment by IljaK91** [comment]
I am also doing it as an exercise :) Will post something in the next few days.

**Comment by IljaK91** [comment]
So far I have come up with this. I am not really a programmer, so I am sure that it needs some changes and that many things could be done in a better way. If I get some feedback, I could fix it also on my own. Would be my first contribution to a package :)


`# Try to split up search requests to get daily Google Trend data
from datetime import datetime, timedelta
from pytrends.request import TrendReq
import pandas as pd
import os
path = ' '
os.chdir(path)
filename = ' '
step = 200
kw_list = [' ',' ']
start_date = datetime(YYYY, MM, DD)

# Login to Google. Only need to run this once, the rest of requests will use the same session.
pytrend = TrendReq()

# Run the first time
today = datetime.today()
old_date = today.date()
# Go 200 days back in time
new_date = today.date() - timedelta(days=step)
# Create new timeframe for which we download data
timeframe = new_date.strftime('%Y-%m-%d')+' '+old_date.strftime('%Y-%m-%d')
pytrend.build_payload(kw_list=kw_list, timeframe = timeframe)
interest_over_time_df = pytrend.interest_over_time()

while new_date>start_date.date():
    # Save the new date from the previous iteration
    old_date = new_date
    # Update the new date to take a step into the past
    new_date = new_date - timedelta(days=step)
    # New timeframe
    timeframe = new_date.strftime('%Y-%m-%d')+' '+old_date.strftime('%Y-%m-%d')
    print(timeframe)
    # Download data
    pytrend.build_payload(kw_list=kw_list, timeframe = timeframe)
    temp_df = pytrend.interest_over_time()
    
    # Renormalize the dataset and drop last line
    for kw in kw_list:
        beg = new_date
        end = old_date - timedelta(days=1)
        if temp_df[kw].iloc[-1] != 0:
            scaling = interest_over_time_df[kw].iloc[0]/temp_df[kw].iloc[-1]
        elif temp_df[kw].iloc[-1] == 0:
            scaling = 0 
    # Basically: if the trend falls to zero, assume that everything before is approximately zero. Not great, but someone should put more thought into this.
        temp_df.loc[beg:end,kw]=temp_df.loc[beg:end,kw]*scaling
    # drop the last row
    temp_df = temp_df[:-1]
    # Cocenate both datasets
    frames = [temp_df,interest_over_time_df]
    interest_over_time_df = pd.concat(frames)
    
# Save dataset
interest_over_time_df.to_csv(filename)

**Comment by dreyco676** [comment]
Awesome. I think you can make a Pull Request. I'll have to refactor this into a function but I think the logic looks sound.

**Comment by IljaK91** [comment]
@dreyco676 Cool! I'll fix some things I wanted to change before creating a Pull Request. :)

**Comment by IljaK91** [comment]
@dreyco676 I created a pull request. I edited yesterday's code quite a bit, such that there is now an overlap of more than just one day. This way we can circumvent the problem of getting a scaling factor for which we need to divide by zero. The idea is still the same. Once you put this in a function and add some things, it should be working.

**Comment by bigme666** [comment]
@IljaK91 as appealing as that sounds, I think this can be made with a straightforward approach in a more integrated fashion. I like your idea and I was looking at the PR to figure it out: I will work on this in the next few days and let you know...

**Comment by maliky** [comment]
Maybe This could help... IljaK91 motivated me to give it a try
```
import pandas as pd
# unofficial API for google trend. !! You need to be connected to google I think
from pytrends.request import TrendReq 
import datetime as dt
import math

#### connect to google
_pytrends = TrendReq(hl='en-US', tz=360)
#### build the playload
_kw_list = ["Blockchain"]  
_cat = 0
_geo = ''
_gprop = '' 
# dates can be `2017-12-07 2018-01-07`, with `today 3-m` `today 5-y`  check trends.google.com's url
_date_fmt = '%Y-%m-%d'
_start_date, _end_date = map(lambda x : dt.datetime.strptime(x, _date_fmt)
                           , ['2017-08-01', '2018-01-05'])

### Building an array of 90d periods to retreive google trend data with a one day resolution
_90d_delta = dt.timedelta(days=90)
_90d_periods = math.ceil( (_end_date - _start_date) / _90d_delta) 

# _tmp_range is a list of date separated by 90d.  We need one more than the number of _90_periods
_tmp_range = pd.date_range(start= _start_date, periods= _90d_periods + 1, freq= '90D')

# making the list of `_start_date _end_date`, strf separated by a space
_rolling_dates = [ ' '.join(map(lambda x : x.strftime(_date_fmt)
                                , [_tmp_range[i], _tmp_range[i+1] ])
                            )
                    for i in range(len(_tmp_range)-1) ]


# _dates contains our last playload argument
# we can now build the final dataframe iterating over the dates
_df_trends = pd.DataFrame()
for _dates in _rolling_dates :
    _pytrends.build_payload(_kw_list
                            , cat=_cat
                            , timeframe=_dates
                            , geo=_geo
                            , gprop=_gprop)
    
    _df_trends = pd.concat([_df_trends, _pytrends.interest_over_time()])
```

**Comment by simonwardjones** [comment]
Hi,

Thanks for the above code.

Just to let you know, I seem to be able to get up to 270 days of data with resolution = 'DAY' for the time series data.

Simon

**Comment by dreyco676** [comment]
Thanks for the work on this all.

**Comment by lopezdp** [comment]
@maliky im getting the following error trying to run this one:

KeyError: 'the label [2009-01-13] is not in the [index]' I just changed the dates to the beginning of the record to present day and that's what I got. Ill try to debug and get you some more feedback.

**Comment by lopezdp** [comment]
> Maybe This could help... @IljaK91 motivated me to give it a try...
> 
> ```python
> import pandas as pd
> from pytrends.request import TrendReq 
> import datetime as dt
> import math
> 
> #### connect to google
> _pytrends = TrendReq(hl='en-US', tz=360)
> #### build the playload
> _kw_list = ["Blockchain"]  
> _cat = 0
> _geo = ''
> _gprop = '' 
> # dates can be formated as  `2017-12-07 2018-01-07`, or  `today 3-m` `today 5-y`  check trends.google.com's url
> _date_fmt = '%Y-%m-%d'
> _start_date, _end_date = map(lambda x : dt.datetime.strptime(x, _date_fmt)
>                            , ['2017-08-01', '2018-01-05'])
> 
> ### Building an array of 90d periods to retreive google trend data with a one day resolution
> _90d_periods = math.ceil( (_end_date - _start_date) / dt.timedelta(days=90) ) 
> 
> # _tmp_range is a list of dates separated by 90d.  We need one more than the number of _90_periods.  if _end_date is in the future google returns the most recent data
> _tmp_range = pd.date_range(start= _start_date, periods= _90d_periods + 1, freq= '90D')
> 
> # making the list of `_start_date _end_date`, strf separated by a space
> _rolling_dates = [ ' '.join(map(lambda x : x.strftime(_date_fmt)
>                                 , [_tmp_range[i], _tmp_range[i+1] ])
>                             )
>                     for i in range(len(_tmp_range)-1) ]
> 
> # initialization of the major data frame _df_trends
> # _dates will contains our last playload argument
> _dates = _rolling_dates[0]
> _pytrends.build_payload(_kw_list, cat=_cat, timeframe=_dates, geo=_geo, gprop=_gprop)
> _df_trends= _pytrends.interest_over_time()
> 
> for _dates in _rolling_dates[1:] :
>     # we need to normalize data before concatanation
>     _common_date = _dates.split(' ')[0]
>     _pytrends.build_payload(_kw_list, cat=_cat, timeframe=_dates, geo=_geo, gprop=_gprop)
>     _tmp_df =   _pytrends.interest_over_time()
>     _multiplication_factor = _df_trends.loc[_common_date] / _tmp_df.loc[_common_date]
> 
>     _df_trends= (pd.concat([_df_trends,
>                            (_tmp_df[1:]* _multiplication_factor)])
>                  .drop(labels = 'isPartial', axis = 1)  # isPartial usefull ?
>                  .resample('D', closed='right').bfill()  # making sure that we have one value per day. 
>                 )
> # _df_trends contains the normalised trends
> ```

Just in case anyone wants to see how I incorporated @maliky implementation you can see it here: http://davidplopez.com/BTC.html You can always drop me a line at davidplopez@live.com also Thank you for this help @maliky
